node/clients/vbase.ts

```
@@ -19,48 +19,91 @@ export default class VBase {
-    await this.client.deleteFile(bucketName(ctx.account), fileName)
+    const testHistory = await this.fetchTestHistory(ctx)
+    testHistory.onGoing = ''
+    testHistory.finishedTests.push(results[0].ABTestBeginning)
+    if (testHistory.finishedTests.length > 100) {
+      testHistory.finishedTests.shift()
+    }
+    try {
+      await this.save(testHistory, abTestHistoryFile, ctx)
+    } catch (ex) {
+      ctx.logger.error(ex)
+      throw new Error(`Something went wrong while finishing the test!`)
+    }
+
+    await this.client.deleteFile(bucketName(ctx.account), testFileName)
     if (results.length > 0) {
       const testResultsFile = 'TestResults' + results[0].ABTestBeginning + '.json'
-      const resultsList = [...await this.fetchResultsList(ctx), testResultsFile]
-      await Promise.all([this.save(results, testResultsFile, ctx), this.save(resultsList, resultsListFile, ctx)])
+      await this.save(results, testResultsFile, ctx)
     }
   }
 
-  private fetchResultsList = async (ctx: IOContext): Promise<string[]> => {
+  private get = async (file: string, ctx: IOContext) => {
     try {
-      const listFile = await this.client.getFile(bucketName(ctx.account), resultsListFile)
-      return JSON.parse(listFile.data.toString()) as string[]
+      const rawFile = await this.client.getFile(bucketName(ctx.account), file)
+      return JSON.parse(rawFile.data.toString())
+    } catch (ex) {
+      ctx.logger.error(ex)
+      throw new Error(`Get request for key ${testFileName} in bucket ${bucketName(ctx.account)} failed!`)
     }
-    catch {
-      return []
+  }
+
+  private fetchTestHistory = async (ctx: IOContext): Promise<ABTestHistory> => {
+    try {
+      const abTestHistory = await this.get(abTestHistoryFile, ctx) as ABTestHistory
+      return abTestHistory
+    } catch (ex) {
+      ctx.logger.error(ex)
+      return {
+        finishedTests: [],
+        onGoing: '',
+      } as ABTestHistory
```

is this a good idea? If something goes wrong and we return this, the original file might me overwritten and all the information lost (this time forever like teardrops in the rain)

---

I was thinking about that also... I accepted this option in face of the tradeoff between this and  getting an error while trying to execute any test operation. But now I changed my mind and do prefer to return an error.